# Course-end Project 1 - EdTech
## by Ernie Aparicio ea@launchoc.com

### Week 1:  Video Segmentation 
#### The first task is to divide the video into keyframes using Uniform Time Sampling.

In [4]:
import cv2
import os

def process_video_for_keyframes(video_path, sampling_rate):
    cap = cv2.VideoCapture(video_path)

    # Check if video opened successfully
    if not cap.isOpened():
        print(f"Error opening video file: {video_path}")
        return

    # Get frame rate
    frame_rate = cap.get(cv2.CAP_PROP_FPS)

    # Calculate sampling interval in frames
    sampling_interval = int(frame_rate * sampling_rate)  # sampling_rate in seconds

    frame_count = 0
    keyframe_count = 0
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        if frame_count % sampling_interval == 0:
            # Folder for keyframes
            base_folder = os.path.splitext(os.path.basename(video_path))[0]
            keyframe_folder = f'keyframes/{base_folder}'
            os.makedirs(keyframe_folder, exist_ok=True)

            # Save keyframe
            keyframe_file = f'{keyframe_folder}/keyframe_{keyframe_count}.jpg'
            cv2.imwrite(keyframe_file, frame)
            keyframe_count += 1

        frame_count += 1

    cap.release()

# List of video paths
video_paths = ['nptel_ai/howToLearn.mp4', 'nptel_ml/ML.mp4']
sampling_rate = 1  # One keyframe every second

# Process each video for keyframes
for video_path in video_paths:
    process_video_for_keyframes(video_path, sampling_rate)

print("Done processing videos.")


Done processing videos.


### Assessment of Instructor Presence and Interaction:
#### We'll use a pre-trained model https://tfhub.dev/tensorflow/ssd_mobilenet_v2 for object detection and check instructor presense.  This model classifies an object as a person.

In [11]:
import os
import cv2
import numpy as np
import tensorflow as tf
import tensorflow_hub as hub

# Load the pre-trained SSD MobileNet model from TensorFlow Hub
model_handle = 'https://tfhub.dev/tensorflow/ssd_mobilenet_v2/fpnlite_640x640/1'
model = hub.load(model_handle).signatures['serving_default']

def detect_instructor(frame, presence_threshold=0.5):
    # Convert the frame to uint8 and process for model input
    frame = tf.image.convert_image_dtype(frame, tf.uint8)
    input_tensor = tf.expand_dims(frame, 0)

    # Model inference
    result = model(input_tensor)

    # Parse the results
    result = {key:value.numpy() for key,value in result.items()}
    detection_scores = result["detection_scores"]
    detection_classes = result["detection_classes"]
    detection_boxes = result["detection_boxes"]

    # Check for instructor presence (class 1 is 'person')
    for score, clss, box in zip(detection_scores[0], detection_classes[0], detection_boxes[0]):
        if score >= presence_threshold and clss == 1:
            return True, box
    return False, None

def is_full_screen(bbox, frame_shape):
    frame_height, frame_width, _ = frame_shape
    bbox_width = bbox[3] - bbox[1]  # Calculate width of bounding box
    bbox_height = bbox[2] - bbox[0]  # Calculate height of bounding box

    # Example thresholds: 50% of frame width and 50% of frame height for full-screen
    min_full_screen_width = frame_width * 0.5
    min_full_screen_height = frame_height * 0.5

    if bbox_width >= min_full_screen_width and bbox_height >= min_full_screen_height:
        return True
    return False

def process_video_segments(segment_folder):
    instructor_presence_count = 0
    total_frames = 0
    full_screen_count = 0
    pip_count = 0

    for segment_file in os.listdir(segment_folder):
        segment_path = os.path.join(segment_folder, segment_file)
        frame = cv2.imread(segment_path)
        if frame is not None:
            total_frames += 1
            present, bbox = detect_instructor(frame)
            if present:
                instructor_presence_count += 1
                if is_full_screen(bbox, frame.shape):
                    full_screen_count += 1
                else:
                    pip_count += 1

    return instructor_presence_count, total_frames, full_screen_count, pip_count

def analyze_video(video_folder):
    instructor_presence, total_frames, full_screen, pip = process_video_segments(video_folder)

    if total_frames > 0:
        fraction_visible = instructor_presence / total_frames
        fraction_full_screen = full_screen / total_frames
        fraction_pip = pip / total_frames
        print(f"Instructor Visibility Fraction in {video_folder}: {fraction_visible}")
        print(f"Fraction of Full-Screen Presence: {fraction_full_screen}")
        print(f"Fraction of PIP Presence: {fraction_pip}")
    else:
        print(f"No frames to analyze in {video_folder}")

# Analyze both videos
analyze_video('keyframes/ML')
analyze_video('keyframes/howToLearn')


Instructor Visibility Fraction in keyframes/ML: 0.9463414634146341
Fraction of Full-Screen Presence: 0.0
Fraction of PIP Presence: 0.9463414634146341
Instructor Visibility Fraction in keyframes/howToLearn: 0.1963882618510158
Fraction of Full-Screen Presence: 0.0
Fraction of PIP Presence: 0.1963882618510158


#### We can see from the results above our model did a good job scoring instructor visibility, full-screen presence, and partial for each video.  In the ML video, majority of keyframes are the instructor and in the How To Learn video, majority of screen is the powerpoint presentation.

#### Next, lets analyze interaction of instructor by using a facial expression model to classify between emotions.  For this I found a pre-trained emotion recognition model from https://github.com/joh-fischer/emotion-recognition-CNN

In [9]:
import os
import cv2
import numpy as np
import tensorflow as tf


# Facial Emotion Recognition Model Setup
from emotion_recognition_model import get_base_model
from emotion_recognition_utils import preprocess_fer, get_labels_fer

IMG_SHAPE = (100, 100, 3)
model = get_base_model(IMG_SHAPE)
model.add(tf.keras.layers.Dense(7, activation='softmax', name="softmax"))

model_name = 'FERplus_0124-1040_weights.h5'
model.load_weights('./models/' + model_name)


def analyze_facial_expression(frame):
    img = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    x = cv2.resize(img, dsize=IMG_SHAPE[:-1])
    x = np.expand_dims(x, axis=0)
    x = preprocess_fer(x)

    output = model.predict(x)
    label = get_labels_fer(output)[0]
    confidence = np.argmax(output[0])

    return label, confidence

def process_keyframes(keyframe_folder):
    expression_counts = {emotion: 0 for emotion in ['neutral', 'happiness', 'surprise', 'sadness', 'anger', 'disgust', 'fear', 'contempt']}
    total_frames = 0

    for keyframe_file in os.listdir(keyframe_folder):
        keyframe_path = os.path.join(keyframe_folder, keyframe_file)
        keyframe = cv2.imread(keyframe_path)
        
        if keyframe is not None:
            total_frames += 1
            label, _ = analyze_facial_expression(keyframe)
            expression_counts[label] += 1

    # Calculate expression percentages
    if total_frames > 0:
        for expression in expression_counts:
            expression_counts[expression] = expression_counts[expression] / total_frames

    return expression_counts

def analyze_videos(video_folders):
    for video_folder in video_folders:
        print(f"Analyzing keyframes in {video_folder}")
        expression_percentages = process_keyframes(video_folder)
        print(f"Facial Expression Percentages for {video_folder}:")
        for expression, percentage in expression_percentages.items():
            print(f"{expression.capitalize()}: {percentage * 100:.2f}%")
        print(f"Finished analyzing {video_folder}")

# List of video keyframe folders to analyze
video_keyframe_folders = ['keyframes/ML', 'keyframes/howToLearn']

# Analyze the videos
analyze_videos(video_keyframe_folders)


Analyzing keyframes in keyframes/ML
1/1 [==============================] - 0s 33ms/step
Facial Expression Percentages for keyframes/ML:
Neutral: 4.39%
Happiness: 69.27%
Surprise: 0.00%
Sadness: 0.00%
Anger: 26.34%
Disgust: 0.00%
Fear: 0.00%
Contempt: 0.00%
Finished analyzing keyframes/ML
Analyzing keyframes in keyframes/howToLearn
1/1 [==============================] - 0s 33ms/step


1/1 [==============================] - 0s 31ms/step


1/1 [==============================] - 0s 33ms/step
Facial Expression Percentages for keyframes/howToLearn:
Neutral: 80.36%
Happiness: 17.61%
Surprise: 0.00%
Sadness: 0.00%
Anger: 2.03%
Disgust: 0.00%
Fear: 0.00%
Contempt: 0.00%
Finished analyzing keyframes/howToLearn


#### Below is a recap of emotional expression for each video. 

#### Facial Expression Percentages for keyframes/ML: Neutral: 4.39% Happiness: 69.27% Surprise: 0.00% Sadness: 0.00% Anger: 26.34% Disgust: 0.00% Fear: 0.00% Contempt: 0.00%

#### Facial Expression Percentages for keyframes/howToLearn: Neutral: 80.36% Happiness: 17.61% Surprise: 0.00% Sadness: 0.00% Anger: 2.03% Disgust: 0.00% Fear: 0.00% Contempt: 0.00%

#### This model reveals that the ML video has an overwhelming happiness score where the howToLearn is more neutral.  This suggests the ML video may provide a more pleasing experience.

#### We're starting to see from our two previous models that the ML video has a happier more visible instructor.  This should lead to greater student experience.  Let's continue.

### Assessment of use of blackboard, slides
#### For this requirement, I want to use ocr to read each frame and determine whether text exists.  If text exists, i want to also detect a person.  If a person, we'll say that frame is a blackboard since the person is interacting with the text if not interacting, we'll say it's a static powerpoint slide.  If no text, than it is neither a blackboard or a powerpoint slide.  We'll use the previous pre trained model for person detection.

In [19]:
import os
import cv2
import tensorflow as tf
import tensorflow_hub as hub
import numpy as np
import pytesseract
from PIL import Image

# Load the pre-trained SSD MobileNet model from TensorFlow Hub
model_handle = 'https://tfhub.dev/tensorflow/ssd_mobilenet_v2/fpnlite_640x640/1'
model = hub.load(model_handle).signatures['serving_default']

def detect_instructor(frame, presence_threshold=0.5):
    # Convert the frame to uint8 and process for model input
    frame = tf.image.convert_image_dtype(frame, tf.uint8)
    input_tensor = tf.expand_dims(frame, 0)

    # Model inference
    result = model(input_tensor)

    # Parse the results
    result = {key:value.numpy() for key,value in result.items()}
    detection_scores = result["detection_scores"]
    detection_classes = result["detection_classes"]
    detection_boxes = result["detection_boxes"]

    # Check for instructor presence (class 1 is 'person')
    for score, clss, box in zip(detection_scores[0], detection_classes[0], detection_boxes[0]):
        if score >= presence_threshold and clss == 1:
            return True, box
    return False, None


def detect_instructor_and_text(frame, presence_threshold=0.5):
    # Detect if an instructor is present in the frame
    instructor_present, instructor_bbox = detect_instructor(frame, presence_threshold)

    # Detect text in the frame using Tesseract OCR
    text_detected = detect_text(frame)

    return instructor_present, instructor_bbox, text_detected


# Function to detect text in the frame using Tesseract OCR
def detect_text(frame):
    # Convert the frame to a PIL Image
    pil_img = Image.fromarray(frame)
    # Use pytesseract to get bounding boxes of text
    boxes = pytesseract.image_to_boxes(pil_img)

    # Calculate the total area covered by text
    text_area = 0
    for b in boxes.splitlines():
        b = b.split(' ')
        text_area += (int(b[3]) - int(b[1])) * (int(b[4]) - int(b[2]))

    frame_area = frame.shape[0] * frame.shape[1]
    text_density = text_area / frame_area if frame_area else 0

    return text_density


def classify_frame(instructor_present, instructor_bbox, text_detected):
    if instructor_present and text_detected:
        return 'blackboard'
    elif text_detected:
        return 'powerpoint'
    else:
        return 'neither'


def process_video_segments(segment_folder):
    # Counters and accumulators
    blackboard_count = 0
    powerpoint_count = 0
    neither_count = 0
    total_frames = 0
    total_text_density = 0

    for segment_file in os.listdir(segment_folder):
        segment_path = os.path.join(segment_folder, segment_file)
        frame = cv2.imread(segment_path)
        if frame is not None:
            total_frames += 1
            person_present, person_bbox, text_density = detect_instructor_and_text(frame)
            classification = classify_frame(person_present, person_bbox, text_density > 0)

            if classification == 'blackboard':
                blackboard_count += 1
            elif classification == 'powerpoint':
                powerpoint_count += 1
            else:
                neither_count += 1

            if text_density > 0:
                total_text_density += text_density

    average_text_density = total_text_density / total_frames if total_frames else 0
    return blackboard_count, powerpoint_count, neither_count, total_frames, average_text_density



def analyze_video(video_folder):
    blackboard, powerpoint, neither, total_frames, avg_text_density = process_video_segments(video_folder)

    if total_frames > 0:
        fraction_blackboard = blackboard / total_frames
        fraction_powerpoint = powerpoint / total_frames
        fraction_neither = neither / total_frames

        print(f"Fraction of Blackboard in {video_folder}: {fraction_blackboard}")
        print(f"Fraction of PowerPoint in {video_folder}: {fraction_powerpoint}")
        print(f"Fraction of Neither in {video_folder}: {fraction_neither}")
        print(f"Average Text Density in each frame {video_folder}: {avg_text_density}")
    else:
        print(f"No frames to analyze in {video_folder}")


# Example usage
analyze_video('keyframes/ML')
analyze_video('keyframes/howToLearn')


Fraction of Blackboard in keyframes/ML: 0.3073170731707317
Fraction of PowerPoint in keyframes/ML: 0.03902439024390244
Fraction of Neither in keyframes/ML: 0.6536585365853659
Average Text Density in each frame keyframes/ML: 0.15745083841463411
Fraction of Blackboard in keyframes/howToLearn: 0.18284424379232506
Fraction of PowerPoint in keyframes/howToLearn: 0.7629796839729119
Fraction of Neither in keyframes/howToLearn: 0.05417607223476298
Average Text Density in each frame keyframes/howToLearn: 0.2583474086448404


#### Our model suggests that the ML course is mostly instructor only.  For the frames with text, the text density is slow. For the howToLearn video, the opposite is true.  Here the instructor relies heavily on powerpoint slides and text density is low.

### Week 2: YouTube comments and reply sentiments detection and Analysis using YouTube API v3
#### Let's get familiar with the Youtube API client.  Below we'll print out some comments.

In [22]:
from googleapiclient.discovery import build
import json

# Replace this with your API key
API_KEY = 'AIzaSyD1FMUZtNeJdcdD9ddmyoV0UMDDFuuTqrM'

# Initialize the YouTube API client
youtube = build('youtube', 'v3', developerKey=API_KEY)

def get_video_comments(video_id):
    # Fetch comments from the video
    comments = []
    response = youtube.commentThreads().list(
        part='snippet,replies',
        videoId=video_id,
        textFormat='plainText',
        maxResults=100  # You can adjust this
    ).execute()

    while response:
        for item in response['items']:
            comment = item['snippet']['topLevelComment']['snippet']['textDisplay']
            comments.append(comment)
            # Check and fetch replies if available
            if 'replies' in item and item['snippet']['totalReplyCount'] > 0:
                for reply_item in item['replies']['comments']:
                    reply = reply_item['snippet']['textDisplay']
                    comments.append(reply)

        # Check if there are more comments
        if 'nextPageToken' in response:
            response = youtube.commentThreads().list(
                part='snippet,replies',
                videoId=video_id,
                pageToken=response['nextPageToken'],
                textFormat='plainText',
                maxResults=100
            ).execute()
        else:
            break

    return comments


# Video IDs
video_ids = ["BRMS3T11Cdw", "zWg7U0OEAoE"]

# Fetch and print comments for each video
for video_id in video_ids:
    print(f"Comments for video {video_id}:")
    comments = get_video_comments(video_id)
    print(json.dumps(comments, indent=2))


Comments for video BRMS3T11Cdw:
[
  "How to join this course",
  "is there is a pdf of  notes of these lecture",
  "Can not find link to register for this course in page",
  "Those who are starting the course\nCheck out these 2 channels for more information regarding ML\n1. statquest by Josh starmer youtube channel\n2. Krish Naik youtube channel",
  "very nice playlist.  Thanks for such course.",
  "Thank you Dr. \ud83e\udd70\ud83e\udd70",
  "Hello madam, very good evening I just want to ask you that is it necessary that only IT person can do this course??",
  "nope, anyone has interest can learn",
  "I am agricultural engineering in machine learning need of corse please reply mam",
  "i am from the mechanical engineering background and have almost no idea of computer programming, can i understand machine learning through this course? anybody please reply",
  "You definitely can. I'd say that you have the very right profile for this. Any Engineering background is a perfect fit for mach

#### Let's categorize the comments (and their replies) as one of the categories: neutral comment, positive comment, negative comment, or a question.  We'll use Textblob.  

In [ ]:
from textblob import TextBlob

def categorize_comment(comment):
    # Analyze the sentiment of the comment
    analysis = TextBlob(comment)
    # Check if it's a question
    if comment.endswith('?'):
        return 'question'
    # Categorize based on sentiment
    if analysis.sentiment.polarity > 0:
        return 'positive comment'
    elif analysis.sentiment.polarity == 0:
        return 'neutral comment'
    else:
        return 'negative comment'

def get_video_comments_with_categories(video_id):
    # Fetch comments from the video
    categorized_comments = []
    response = youtube.commentThreads().list(
        part='snippet,replies',
        videoId=video_id,
        textFormat='plainText',
        maxResults=100  # You can adjust this
    ).execute()

    while response:
        for item in response['items']:
            comment = item['snippet']['topLevelComment']['snippet']['textDisplay']
            category = categorize_comment(comment)
            categorized_comments.append((comment, category))
            # Check and fetch replies if available
            if 'replies' in item and item['snippet']['totalReplyCount'] > 0:
                for reply_item in item['replies']['comments']:
                    reply = reply_item['snippet']['textDisplay']
                    reply_category = categorize_comment(reply)
                    categorized_comments.append((reply, reply_category))

        # Check if there are more comments
        if 'nextPageToken' in response:
            response = youtube.commentThreads().list(
                part='snippet,replies',
                videoId=video_id,
                pageToken=response['nextPageToken'],
                textFormat='plainText',
                maxResults=100
            ).execute()
        else:
            break

    return categorized_comments

# Fetch and categorize comments for each video
for video_id in video_ids:
    print(f"Categorized comments for video {video_id}:")
    categorized_comments = get_video_comments_with_categories(video_id)
    for comment, category in categorized_comments:
        print(f"Comment: {comment}\nCategory: {category}\n")
